src/click/shell_completion.py

```diff
@@ -0,0 +1,468 @@
+    completion_object = completion_source(prog_name, complete_var, cli)
+
+    return (
+        completion_object.source()
+        % {
+            "complete_func": f"_{cf_name}_completion",
+            "script_names": prog_name,
+            "autocomplete_var": complete_var,
+        }
+    ).strip() + ";"
+
+
+def add_completion_class(name, completion_class):
+    """Adds the given completion class to the dictionary of available
+    shells under the given name. If the completion class does not
+    extend :class:`ShellComplete`, an error is raised.
+
+    :param name: the name of the shell. Must match the shell name used
+                 during activation: `source_{shell}`.
+    :param completion_class: the class to be instantiated for completion
+    """
+    global available_shells
+
+    if not issubclass(completion_class, ShellComplete):
+        raise TypeError("Completion class must extend 'ShellComplete'")
+
+    available_shells[name] = completion_class
+
+
+def get_completion_source(shell):
+    """Returns the completion class using the given shell as the key, and
+    defaults to the `ShellComplete` class otherwise.
+
+    :param shell: the current shell
+    """
+    return available_shells.get(shell, ShellComplete)
+
+
+def is_incomplete_argument(current_params, cmd_param):
+    """Returns whether or not the last argument is complete and corresponds
+    to `cmd_param`. In other words, determine if `cmd_param` can still accept
+    values.
+
+    :param current_params: the current params and values for this argument
+                           as already entered
+    :param cmd_param: the current command parameter
+    """
+    if not isinstance(cmd_param, Argument):
+        return False
+    current_param_values = current_params[cmd_param.name]
+    if current_param_values is None:
+        return True
+    if cmd_param.nargs == -1:
+        return True
+    if (
+        isinstance(current_param_values, abc.Iterable)
+        and cmd_param.nargs > 1
+        and len(current_param_values) < cmd_param.nargs
+    ):
+        return True
+    return False
+
+
+def is_incomplete_option(all_args, cmd_param):
+    """Returns whether or not the last option declaration (i.e. starts with
+    "-" or "--") is incomplete and corresponse to `cmd_param`. In other words,
+    determine if `cmd_param` can still accept values.
+
+    :param all_args: the full original list of args supplied
+    :param cmd_param: the current command parameter
+    """
+    if not isinstance(cmd_param, Option):
+        return False
+    if cmd_param.is_flag:
+        return False
+    last_option = None
+    for index, arg_str in enumerate(
+        reversed([arg for arg in all_args if arg != WORDBREAK])
+    ):
+        if index + 1 > cmd_param.nargs:
+            break
+        if start_of_option(arg_str):
+            last_option = arg_str
+
+    return True if last_option and last_option in cmd_param.opts else False
```

Comment:
return true if true, else return false? 
could this be simplified by `return last_option and last_option in cmd_param.opts`?

Reply:
Yep, nice catch!